enhancements/kube-apiserver/certificates.md

```diff
@@ -0,0 +1,144 @@
+
+## Summary
+
+This is documentation about the certificates that Kubernetes has and how the certificates
+for the kube-apiserver in particular are managed.
+
+## Motivation
+
+We face many questions about 
+1. What a specific certificate/flag is for.
+2. How a specific certificate is created.
+3. What breaks when certificate fails.
+4. How often a certificate is rotate.
+
+This document attempts to answer these questions.
+
+
+### Goals
+
+### Non-Goals
+
+## Proposal
+
+### TLS related flags for individual binaries
+
+#### kube-apiserver
+| Flag | What’s it for |
+|---|---|
+| `--etcd-cafile`  | CA bundle used to verify the etcd server really is the etcd server |
+| `--etcd-certfile` | Client cert used to identify KAS to the etcd server |
+| `--etcd-keyfile` | Client key used to identify KAS to the etcd server |
+| `--tls-cert-file` | Serving cert used to serve requests not matching SNI.  Must be verifiable with `kube-controller-manager --root-ca-file` or service must be handled with SNI. |
+| `--tls-private-key-file` | Serving key used to serve requests not matching SNI |
+| `--tls-sni-cert-key` | Special flag format to specify hostname-pattern,cert,key tuples to serve matching SNI requests.  If used for kubernetes.default.service, must be verifiable with `kube-controller-manager --root-ca-file`.. |
+| `--client-ca-file` | CA bundle used to verify client certificate connections from clients and identify users. (I am Bob).  Must be able to verify `kube-controller-manager --cluster-signing-cert-file` or `kubelet --rotate-certificates` will fail. |
+| `--requestheader-client-ca-file` | CA bundle used to verify client certificate connections from front proxies that are asserting the identity of user. (This request is from Bob).  Must be able to verify `kube-apiserver --proxy-client-cert-file` or aggregation in the cluster will fail by default. |
+| `--kubelet-certificate-authority` | CA bundle used to verify kubelets for connections from KAS to kubelet.  (Think logs,exec,etc).  Must be able to verify `kubelet --tls-cert-file`.  Must be able to verify `kube-controller-manager --cluster-signing-cert-file` or `kubelet --rotate-server-certificates` will fail. |
+| `--kubelet-client-certificate` | Client cert used to identify KAS to the kubelets.  Must be verifiable by `kubelet --client-ca-file`. |
+| `--kubelet-client-key` | Client key used to identify KAS to the kubelets |
+| `--proxy-client-cert-file` | Client cert used to identify KAS to aggregated API servers as a front proxy.  Must be verifiable by `kube-apiserver --requestheader-client-ca-file` or aggregation in the cluster will fail by default |
+| `--proxy-client-key-file` | Client key used to identify KAS to aggregated API servers as a front proxy |
+| `--service-account-key-file` | RSA keys used to verify ServiceAccount tokens.  Must be able to verify `kube-controller-manager --service-account-private-key-file` for all keys you want to continue working. |
```

Comment:
`service-account-signing-key-file` will be coming soon too.  Also maybe mention that this is hard to rotate "completely" since SA tokens are long lived (i.e. even if you rotate to a new signing key you need to keep old public keys around).

Reply:
> `service-account-signing-key-file` will be coming soon too. Also maybe mention that this is hard to rotate "completely" since SA tokens are long lived (i.e. even if you rotate to a new signing key you need to keep old public keys around).

you will get to open the enhancment when it comes :)